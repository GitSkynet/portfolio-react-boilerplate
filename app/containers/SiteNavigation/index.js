import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import {
  makeSelectLocation,
  makeSelectSiteNavigationIsAtScreenTop,
  makeSelectSiteWidth,
} from '../../containers/App/selectors';
import {
  setSiteNavigationOffsetHeight,
  setSiteNavigationTopPosition,
} from '../../containers/App/actions';

import Navigation from '../../components/Navigation';

import getElementTop from '../../utils/getElementTop';

class SiteNavigation extends React.PureComponent {
  constructor(props) {
    super(props);

    const IS_BIG_SCREEN = (this.props.siteWidth >= this.props.bigScreenBreakpoint);

    this.state = {
      isBigScreen: IS_BIG_SCREEN,
      isExpanded: IS_BIG_SCREEN,
    };

    this.siteNavigation = React.createRef();
    this.toggleMenu = this.toggleMenu.bind(this);
  }

  componentDidMount() {
    const myTopPosition = getElementTop(this.siteNavigation) || 0;
    const myOffsetHeight = (this.siteNavigation) ? this.siteNavigation.offsetHeight : 0;
    this.props.setTopPosition(myTopPosition);
    this.props.setOffsetHeight(myOffsetHeight);
  }

  componentDidUpdate(prevProps) {
    if (this.props.siteWidth !== prevProps.siteWidth) {
      const IS_BIG_SCREEN = (this.props.siteWidth >= this.props.bigScreenBreakpoint);
      this.setState({ // eslint-disable-line react/no-did-update-set-state
        isBigScreen: IS_BIG_SCREEN,
        isExpanded: IS_BIG_SCREEN,
      });
    }
  }

  componentWillUnmount() {
    this.props.setTopPosition(0);
    this.props.setOffsetHeight(0);
  }

  toggleMenu() {
    this.setState((prevState) => ({
      isExpanded: !prevState.isExpanded,
    }));
  }

  render() {
    return (
      <Navigation
        bigScreenBreakpoint={this.props.bigScreenBreakpoint}
        isAtScreenTop={this.props.isAtScreenTop || false}
        navigationRef={(el) => { this.siteNavigation = el; }}
        isBigScreen={this.state.isBigScreen}
        isExpanded={this.state.isExpanded}
        siteWidth={this.props.siteWidth}
        toggleMenu={this.toggleMenu}
      />
    );
  }
}

SiteNavigation.propTypes = {
  bigScreenBreakpoint: PropTypes.number, // PIXELS!
  setTopPosition: PropTypes.func.isRequired,
  setOffsetHeight: PropTypes.func.isRequired,
  isAtScreenTop: PropTypes.bool,
  siteWidth: PropTypes.number,
};

SiteNavigation.defaultProps = {
  bigScreenBreakpoint: 1600,
};

export function mapDispatchToProps(dispatch) {
  return {
    setTopPosition: (topPosition) => (dispatch(setSiteNavigationTopPosition(topPosition))),
    setOffsetHeight: (offsetHeight) => (dispatch(setSiteNavigationOffsetHeight(offsetHeight))),
  };
}

const mapStateToProps = createStructuredSelector({
  location: makeSelectLocation(),
  isAtScreenTop: makeSelectSiteNavigationIsAtScreenTop(),
  siteWidth: makeSelectSiteWidth(),
});

export default connect(mapStateToProps, mapDispatchToProps)(SiteNavigation);
