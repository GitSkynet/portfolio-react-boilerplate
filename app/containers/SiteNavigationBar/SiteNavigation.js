import PropTypes from 'prop-types';
import styled from 'styled-components';
import { rem } from 'polished';

const SiteNavigation = styled.nav`
  width: 100%;
  padding: 0.35rem 0;
  position: ${(props) => props.isAtScreenTop ? 'fixed' : 'unset'};
  top: 0;
  z-index: 9999;
  background-color: #575756;

  @media (min-width: 600px) {
    left: 100%;
    position: ${(props) => props.isAtScreenTop ? 'fixed' : 'absolute'};
    top: ${(props) => props.isAtScreenTop ? 0 : rem('77px')};
    width: unset;
    padding: 0 0.35rem;
    transform: rotate(90deg);
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;
  }
`;

SiteNavigation.propTypes = {
  isAtScreenTop: PropTypes.bool,
};

export default SiteNavigation;
