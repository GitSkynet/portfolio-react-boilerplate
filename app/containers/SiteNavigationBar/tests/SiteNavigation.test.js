import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import SiteNavigation from '../SiteNavigation';

describe('<SiteNavigation />', () => {
  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(<SiteNavigation />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "isAtScreenTop" prop', () => {
    const isAtScreenTop = false;
    const renderedComponent = mount(<SiteNavigation isAtScreenTop={isAtScreenTop} />);
    expect(renderedComponent.prop('isAtScreenTop')).toBe(isAtScreenTop);
  });
});
