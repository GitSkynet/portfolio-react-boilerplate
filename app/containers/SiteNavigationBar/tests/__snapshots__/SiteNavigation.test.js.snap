// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SiteNavigation /> should render and match the snapshot 1`] = `
.c0 {
  width: 100%;
  padding: 0.35rem 0;
  position: unset;
  top: 0;
  z-index: 9999;
  background-color: #575756;
  box-shadow: 0 3px 5px rgba(0,0,0,0.3);
}

@media (min-width:600px) {
  .c0 {
    left: 100%;
    position: absolute;
    top: 4.8125rem;
    width: unset;
    padding: 0 0.35rem;
    -webkit-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
    -webkit-transform-origin: top left;
    -ms-transform-origin: top left;
    transform-origin: top left;
    background-color: transparent;
    border-bottom: none;
    box-shadow: none;
  }
}

<nav
  className="c0"
/>
`;
